### PR TITLE
General brightness improvements: new `brightnessctl` backend, set by percentage, and pick device

### LINKED ISSUE #27: Control brightness of all available devices
Hi, I am using external monitors and was wondering if `swayosd --brightness` could control all devices in `/sys/class/backlight`?

Thanks a lot!

### LINKED ISSUE #37: Feature request: Raise and lower by a percentage
Hi,

Great work on SwayOSD!

Both `pactl` and `brightnessctl` offer the possibility to change volume and brightness by a percentage, which is supposed to be [closer to how humans perceive changes](https://medium.com/zattoo_tech/logarithmic-volume-control-bd89cc1f2135).

Would you consider making this possible with SwayOSD also? It would also help with compatibility I guess.

### LINKED ISSUE #56: Less intrusive brightness control (and other?)
I, for one, am less fond of messing with groups and udev rules to change brightness when I already have other utilities that can do that without such intrusions (e.g. `brightnessctl`). I would appreciate a design that relies less on lower-level APIs and/or just having options similar to the ones for caps lock on the 'continuous' inputs as well (rather than the hack of setting to +0), so I can compose my preferred control methods appropriately.

### PR DESCRIPTION
**Pull request by Syndelis**
Resolves #56 
Resolves #37 
*Partially* Resolves #27 

## About this PR

This PR separates the code that controls brightness into a new module. This module exports a trait that different backends can implement in order to control brightness like SwayOSD expects.

Besides the existing "blight" backend, a new backend has been introduced, one that is *backed* by the [command line util `brightnessctl`](https://github.com/Hummer12007/brightnessctl) (#56). This backend is attempted first and "blight" serves as a fallback if the command is not available. I've made it this way because "blight" requires some setting up, while `brightnessctl` requires only the command to be installed.

The brightness command also accepts `brightness={val}` to set the brightness directly in percentages (#37).

Finally, the PR also allows for the user to pick a different device to have its brightness set, via the already present `--device` parameter. I believe this partially resolves #27 because the user could now make a script to call SwayOSD for each device, which was previously impossible.

### CODE REVIEW

**Comment by Syndelis** [comment]
Hi there, @ErikReider ! Can I get some feedback on this one? I'll be happy to make any adjustments!

File: src/brightness_backend/brightnessctl.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        let s = Self { name: device_name.clone(), ..Default::default() };
+
+        // Check if the command is available to us before running it in other occasions
+        let exit_code = s.command(CliArg::Simple("info"))
+            .output()?
+            .status;
+
+        if exit_code.success() {
+            Ok(s)
+        }
```
Comment (ErikReider) [inline src/brightness_backend/brightnessctl.rs:57]: Any other way of doing this without calling the executable? Does brightnessctl have any APIs that can be used?

**Review by ErikReider — CHANGES_REQUESTED** [review]
Please excuse me for the late review 😅

File: src/brightness_backend/brightnessctl.rs (hunk 1 above)
Comment (Syndelis) [reply src/brightness_backend/brightnessctl.rs:57]: Not really, [brightnessctl](https://github.com/Hummer12007/brightnessctl) is, in fact, made in a single `.c` file

**Review by ErikReider — CHANGES_REQUESTED** [review]
Other than that, it LGTM :)

File: src/brightness_backend/mod.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+pub trait BrightnessBackend {
+    fn get_current(&mut self) -> u32;
+    fn get_max(&mut self) -> u32;
+
+    fn lower(&mut self, by: u32) -> anyhow::Result<()>;
+    fn raise(&mut self, by: u32) -> anyhow::Result<()>;
+    fn set(&mut self, val: u32) -> anyhow::Result<()>;
+}
+
+pub fn get_preferred_backend(device_name: Option<String>) -> BrightnessBackendResult {
```
[the remark this answers is not in the document]
Comment (Syndelis) [reply src/brightness_backend/mod.rs:31]: I honestly do not know why it thinks the function is unused. Added `#[allow(dead_code)]` in dc07963

File: src/brightness_backend/mod.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    println!("Trying BrightnessCtl Backend...");
+    BrightnessCtl::try_new_boxed(device_name.clone()).or_else(|_| {
+        println!("...Command failed! Falling back to Blight");
+        Blight::try_new_boxed(device_name)
+    })
+}
```
[the remark this answers is not in the document]
Comment (Syndelis) [reply src/brightness_backend/mod.rs]: Addressed in ba06c46

**Comment by Syndelis** [comment]
@ErikReider I've addressed the review comments!

**Review by ErikReider — APPROVED** [review]
Thanks for the amazing work! :)

**Pull request merged**